### PR TITLE
Add feature flags for Keeper API

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Keeper improvement: add feature flags for Keeper API. Each feature flag can be disabled or enabled by defining it under `keeper_server.feature_flags` config. E.g. to enable `CheckNotExists` request, `keeper_server.feature_flags.check_not_exists` should be set to `1` on Keeper

cc @nikitamikhaylov @alesapin 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 07582d56f32d72a3f13d9e7303310ff1753c97f5 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50796/07582d56f32d72a3f13d9e7303310ff1753c97f5/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Coordination/KeeperContext.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+DECLARE_SETTING_ENUM(KeeperApiVersion);
+IMPLEMENT_SETTING_ENUM(KeeperApiVersion, ErrorCodes::BAD_ARGUMENTS,
+    {{"ZOOKEEPER_COMPATIBLE", KeeperApiVersion::ZOOKEEPER_COMPATIBLE},
+     {"WITH_FILTERED_LIST", KeeperApiVersion::WITH_FILTERED_LIST},
+     {"WITH_MULTI_READ", KeeperApiVersion::WITH_MULTI_READ},
+     {"WITH_CHECK_NOT_EXISTS", KeeperApiVersion::WITH_CHECK_NOT_EXISTS}});
+
+void KeeperContext::initialize(const Poco::Util::AbstractConfiguration & config)
```
Comment (antonio2368) [inline src/Coordination/KeeperContext.cpp:23]: Larger refactor of `KeeperContext` is in https://github.com/ClickHouse/ClickHouse/pull/50098 so I didn't modify it much in this PR

**Comment by nikitamikhaylov** [comment]
It seems that `ZOOKEEPER_COMPATIBLE` < `WITH_FILTERED_LIST` < `WITH_MULTI_READ` < `WITH_CHECK_NOT_EXISTS`, but how should I know this without looking into the code?

**Comment by antonio2368** [comment]
@nikitamikhaylov you think it's better to just use number?

**Comment by nikitamikhaylov** [comment]
Ideal scenario would be to implement several independent settings Like 
- use_filtered_list
- use_multi_read
...

But it breaks compatibility... So, some increasing "version" number would be better I think

**Comment by alexey-milovidov** [comment]
Please take a look at the broken tests.

**Comment by antonio2368** [comment]
@nikitamikhaylov I updated PR to use feature flags instead so we can enable/disable any feature we would like
Need to check how it behaves with different older versions but this should be more or less it.

**Comment by nikitamikhaylov** [comment]
@antonio2368 Perfect! Aaand I have another weird idea - to implement another 4LW `gftr` to output a list of all the features that could be enabled...

**Comment by antonio2368** [comment]
> to implement another 4LW gftr to output a list of all the features that could be enabled...

`ftfl` already returns list of all features with information if its enabled

```
check_not_exists 0
multi_read 1
```

File: src/Storages/System/StorageSystemZooKeeperConnection.cpp [hunk 2]
```diff
@@ -31,7 +31,7 @@ void StorageSystemZooKeeperConnection::fillData(MutableColumns & res_columns, Co
     res_columns[3]->insert(context->getZooKeeper()->getConnectedZooKeeperIndex());
     res_columns[4]->insert(context->getZooKeeperSessionUptime());
     res_columns[5]->insert(context->getZooKeeper()->expired());
-    res_columns[6]->insert(context->getZooKeeper()->getApiVersion());
```
Comment (nikitamikhaylov) [inline src/Storages/System/StorageSystemZooKeeperConnection.cpp:34]: @antonio2368 Probably we can provide a list of features that can be used in a scope of this connection instead?

File: src/Coordination/KeeperFeatureFlags.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#include <Coordination/KeeperConstants.h>
+
+namespace DB
+{
+
+enum KeeperFeatureFlag
+{
+    FILTERED_LIST = 0,
+    MULTI_READ,
+    CHECK_NOT_EXISTS,
```
Comment (nikitamikhaylov) [inline src/Coordination/KeeperFeatureFlags.h:14]: A comment is needed that you can't reorder these values - only add.

File: src/Coordination/KeeperFeatureFlags.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <Core/SettingsEnums.h>
+#include <Core/SettingsFields.h>
+#include <Coordination/KeeperConstants.h>
+
+namespace DB
+{
+
+enum KeeperFeatureFlag
```
Comment (nikitamikhaylov) [inline src/Coordination/KeeperFeatureFlags.h:10]:
```suggestion
enum KeeperFeatureFlag : size_t 
```

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by lzydmxy** [comment]
@antonio2368 Hi, can we merge it into the previous version(23.3)? In other words, I'm now using version 23.3, and zookeeper, and I want to use the multiread feature

**Comment by antonio2368** [comment]
@lzydmxy sorry we don't backport features and improvements only important bugfixes
Also, this feature will only work with Keeper's that support MultiRead because it's slightly different from ZKs MultiRead feature.

**Comment by lzydmxy** [comment]
> @lzydmxy sorry we don't backport features and improvements only important bugfixes Also, this feature will only work with Keeper's that support MultiRead because it's slightly different from ZKs MultiRead feature.
Okay, I get it. The difference is Keeper support more read ops in MultiRead?(For example:FilteredList)

**Comment by antonio2368** [comment]
> Okay, I get it. The difference is Keeper support more read ops in MultiRead?(For example:FilteredList)

Exactly + we can extend more easily because CH internal ZK client automatically detects which features Keeper it connected to supports.

**Comment by lzydmxy** [comment]
@antonio2368 Thank you for your answer  :)

**Pull request merged**